<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
box-sizing: border-box;
min-width: 200px;
max-width: 980px;
margin: 0 auto;
padding: 45px;
padding-top: 0px;
}
</style>


</head>

<body>

<!-- README.md is generated from README.Rmd. Please edit that file -->

<h1 id="icikendalltau">ICIKendallTau</h1>
<!-- badges: start -->

<p><a href="https://moseleybioinformaticslab.r-universe.dev"><img role="img" aria-label="ICIKendallTau status badge" src="data:image/svg+xml; charset=utf-8;base64,PHN2ZyB3aWR0aD0iMTEyLjgiIGhlaWdodD0iMjAiIHZpZXdCb3g9IjAgMCAxMTI4IDIwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiByb2xlPSJpbWciIGFyaWEtbGFiZWw9InItdW5pdmVyc2U6IDEuMi4xMCI+CiAgPGEgaHJlZj0iaHR0cHM6Ly9tb3NlbGV5YmlvaW5mb3JtYXRpY3NsYWIuci11bml2ZXJzZS5kZXYvSUNJS2VuZGFsbFRhdSIgYWx0PSJyLXVuaXZlcnNlIj4KICA8dGl0bGU+ci11bml2ZXJzZTogMS4yLjEwPC90aXRsZT4KICA8bGluZWFyR3JhZGllbnQgaWQ9InByUUlBIiB4Mj0iMCIgeTI9IjEwMCUiPgogICAgPHN0b3Agb2Zmc2V0PSIwIiBzdG9wLW9wYWNpdHk9Ii4xIiBzdG9wLWNvbG9yPSIjRUVFIi8+CiAgICA8c3RvcCBvZmZzZXQ9IjEiIHN0b3Atb3BhY2l0eT0iLjEiLz4KICA8L2xpbmVhckdyYWRpZW50PgogIDxtYXNrIGlkPSJLamZPRiI+PHJlY3Qgd2lkdGg9IjExMjgiIGhlaWdodD0iMjAwIiByeD0iMzAiIGZpbGw9IiNGRkYiLz48L21hc2s+CiAgPGcgbWFzaz0idXJsKCNLamZPRikiPgogICAgPHJlY3Qgd2lkdGg9IjY2OCIgaGVpZ2h0PSIyMDAiIGZpbGw9IiM1NTUiLz4KICAgIDxyZWN0IHdpZHRoPSI0NjAiIGhlaWdodD0iMjAwIiBmaWxsPSIjM0MxIiB4PSI2NjgiLz4KICAgIDxyZWN0IHdpZHRoPSIxMTI4IiBoZWlnaHQ9IjIwMCIgZmlsbD0idXJsKCNwclFJQSkiLz4KICA8L2c+CiAgPGcgYXJpYS1oaWRkZW49InRydWUiIGZpbGw9IiNmZmYiIHRleHQtYW5jaG9yPSJzdGFydCIgZm9udC1mYW1pbHk9IlZlcmRhbmEsRGVqYVZ1IFNhbnMsc2Fucy1zZXJpZiIgZm9udC1zaXplPSIxMTAiPgogICAgPHRleHQgeD0iNjAiIHk9IjE0OCIgdGV4dExlbmd0aD0iNTY4IiBmaWxsPSIjMDAwIiBvcGFjaXR5PSIwLjI1Ij5yLXVuaXZlcnNlPC90ZXh0PgogICAgPHRleHQgeD0iNTAiIHk9IjEzOCIgdGV4dExlbmd0aD0iNTY4Ij5yLXVuaXZlcnNlPC90ZXh0PgogICAgPHRleHQgeD0iNzIzIiB5PSIxNDgiIHRleHRMZW5ndGg9IjM2MCIgZmlsbD0iIzAwMCIgb3BhY2l0eT0iMC4yNSI+MS4yLjEwPC90ZXh0PgogICAgPHRleHQgeD0iNzEzIiB5PSIxMzgiIHRleHRMZW5ndGg9IjM2MCI+MS4yLjEwPC90ZXh0PgogIDwvZz4KICAKICA8L2E+Cjwvc3ZnPg==" alt="ICIKendallTau status badge" /></a></p>
<!-- badges: end -->

<p>You can see the pkgdown site <a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/">here</a>.</p>
<h2 id="citation">Citation</h2>
<p>This package has an <a href="https://doi.org/10.1101/2022.02.24.481854">associated preprint on
bioRxiv</a>, please cite it if you use the package in your own work.</p>
<pre><code>Flight RM, Bhatt PS, Moseley HNB (2025). “Information-Content-Informed Kendall-tau Correlation Methodology: Interpreting Missing Values as Useful Information.”
_bioRxiv_. doi:10.1101/2022.02.24.481854 </code></pre>
<h2 id="installation">Installation</h2>
<p>You can install the current version of ICIKendallTau via GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;MoseleyBioinformaticsLab/ICIKendallTau&quot;</span>)</span></code></pre></div>
<p>You can also install Windows or Mac binaries using our
r-universe:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repos =</span> <span class="fu">c</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>    <span class="at">moseleybioinformaticslab =</span> <span class="st">&#39;https://moseleybioinformaticslab.r-universe.dev&#39;</span>,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    <span class="at">CRAN =</span> <span class="st">&quot;https://cloud.r-project.org&quot;</span>))</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;ICIKendallTau&quot;</span>)</span></code></pre></div>
<h2 id="problem">Problem</h2>
<ul>
<li>How to handle missing data (i.e. <code>NA</code>’s) in calculating a
correlation between two variables.</li>
<li>Current calculations of correlation are based on having all pairs of
observations for two variables.
<ul>
<li>However, whether an observation is present or missing is
semi-quantitative information for many analytical measurements with
sensitivity limits.</li>
<li>i.e. in many cases, missing observations are not
“missing-at-random”, but “missing-not-at-random” due to falling below
the detection limit.</li>
<li>In these cases, NA is informative.</li>
<li>Therefore, in <strong>most</strong> analytical measurements (gene
expression, proteomics, metabolomics), missing measurements should be
included, and contribute to the correlation.</li>
</ul></li>
</ul>
<p>If you want to read more on <strong>how</strong> we solve this
problem, see the package <a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/articles/ici-kendalltau.html">vignette</a>.</p>
<h2 id="package-functions">Package Functions</h2>
<p>The functions that implement this include:</p>
<ul>
<li><code>ici_kt</code>: the C++ workhorse, actually calculating a
correlation between an X and Y.
<ul>
<li>The option <code>perspective</code> will control how the
<code>NA</code> values influence ties.</li>
<li>When comparing samples, you likely want to use
<code>perspective = &quot;global&quot;</code>.</li>
</ul></li>
<li><code>ici_kendallt</code>: Handles comparisons for a large matrix.
<ul>
<li>Rows are features, columns are samples.</li>
<li>Implicitly parallel, but have to call:
<ul>
<li><code>library(furrr)</code></li>
<li><code>plan(multiprocess)</code></li>
</ul></li>
<li>Otherwise will only use a single core.</li>
</ul></li>
</ul>
<p>We’ve also included a function for testing if the missingness in your
data comes from left-censorship, <code>test_left_censorship</code>. We
walk through creating example data and testing it in the vignette <a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/articles/testing-for-left-censorship">Testing
for Left Censorship</a>. In addition to testing, you can also visualize
the missing data pattern by feature rank using the
<code>rank_order_data</code> function, and use
<code>visdat::vis_miss()</code> on the original and reordered missing
data.</p>
<h2 id="examples">Examples</h2>
<p>The most common case is a large matrix of independent samples
(columns) and measured features in each of the samples (i.e. gene
expression).</p>
<p>Here we will make some artificial data to show how the correlation
changes as we introduce missingness.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">library</span>(ICIKendallTau)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>s1 <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">rnorm</span>(<span class="dv">1000</span>, <span class="at">mean =</span> <span class="dv">100</span>, <span class="at">sd =</span> <span class="dv">10</span>))</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>s2 <span class="ot">=</span> s1 <span class="sc">+</span> <span class="dv">10</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>matrix_1 <span class="ot">=</span> <span class="fu">cbind</span>(s1, s2)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>r_1 <span class="ot">=</span> <span class="fu">ici_kendalltau</span>(matrix_1)</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>r_1<span class="sc">$</span>cor</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt;    s1 s2</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; s1  1  1</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt; s2  1  1</span></span></code></pre></div>
<p>Now we introduce some missing values at the low end of each one. We
will just do the simplest thing and introduce <code>NA</code> values in
the bottom set.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>s3 <span class="ot">=</span> s1</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>s3[<span class="fu">sample</span>(<span class="dv">100</span>, <span class="dv">50</span>)] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>s4 <span class="ot">=</span> s2</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>s4[<span class="fu">sample</span>(<span class="dv">100</span>, <span class="dv">50</span>)] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>matrix_2 <span class="ot">=</span> <span class="fu">cbind</span>(s3, s4)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>r_2 <span class="ot">=</span> <span class="fu">ici_kendalltau</span>(matrix_2)</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>r_2<span class="sc">$</span>cor</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt;           s3        s4</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; s3 1.0000000 0.9944616</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; s4 0.9944616 1.0000000</span></span></code></pre></div>
<h2 id="is-it-fast">Is It Fast?</h2>
<p>The C++ code implementation (thanks {Rcpp}!) is based on the SciPy
implementation, which uses two merge sorts of the ranks of each vector,
and then looks for differences between them. This is the fastest method
we know of, and has a complexity of O(nlogn). The naive way of computing
it, which explicitly examines all of the pairs, has a complexity of n^2.
Our implementation was compared to the {pcaPP::cov.fk} function, and the
use of {Rcpp} and our inefficient copying of vectors makes ours 3X
slower than that one. Which honestly isn’t too bad.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">library</span>(microbenchmark)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>x2 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="dv">40000</span>)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>y2 <span class="ot">=</span> <span class="fu">rnorm</span>(<span class="dv">40000</span>)</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="fu">microbenchmark</span>(</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  <span class="fu">cor</span>(x, y, <span class="at">method =</span> <span class="st">&quot;kendall&quot;</span>),</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>  <span class="fu">ici_kt</span>(x, y, <span class="st">&quot;global&quot;</span>),</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>  <span class="fu">ici_kt</span>(x2, y2, <span class="st">&quot;global&quot;</span>),</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>  <span class="at">times =</span> <span class="dv">5</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>)</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt; Unit: microseconds</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt;                           expr       min        lq      mean    median        uq       max neval</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt;  cor(x, y, method = &quot;kendall&quot;) 12020.945 12227.957 12778.679 12241.762 13459.533 13943.197     5</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt;         ici_kt(x, y, &quot;global&quot;)   331.001   346.866   456.299   361.061   401.216   841.351     5</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt;       ici_kt(x2, y2, &quot;global&quot;) 18462.550 18468.493 20342.581 19215.913 22400.049 23165.899     5</span></span></code></pre></div>
<p>In the case of 40,000 features, the average time on a modern CPU is
14 milliseconds.</p>
<p>Of course, if you want to use it to calculate Kendall-tau-b without
incorporating missingness, it can do that just fine as well.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>k_tau <span class="ot">=</span> <span class="fu">ici_kt</span>(x, y, <span class="st">&quot;global&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">all.equal</span>(k_tau[[<span class="dv">1</span>]], <span class="fu">cor</span>(x, y, <span class="at">method =</span> <span class="st">&quot;kendall&quot;</span>))</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>We also provide the <code>kt_fast</code> function, if you want
something that treats <code>NA</code> values similarly to
<code>stats::cor</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>k_tau_fast <span class="ot">=</span> <span class="fu">kt_fast</span>(x, y)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>k_tau_fast</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#&gt; $tau</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt;              x            y</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; x  1.000000000 -0.003411411</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; y -0.003411411  1.000000000</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; $pvalue</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt;           x         y</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; x 0.0000000 0.8716723</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; y 0.8716723 0.0000000</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; $run_time</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; [1] 0.02020574</span></span></code></pre></div>
<h2 id="p-values">P-Values</h2>
<p>ICI-Kt functions only calculates the tau-b variant that handles ties.
P-value calculations use the asymptotic approximation in all cases, and
thus may vary slightly from the p-values returned by R’s
<code>cor.test</code> and Python’s <code>scipy.stats.kendalltau</code>
depending on the number of values in <em>x</em> and <em>y</em>.</p>
<h2 id="parallelism">Parallelism</h2>
<p>If you have {future} and the {furrr} packages installed, then it is
also possible to split up the a set of matrix comparisons across compute
resources for any multiprocessing engine registered with {future}.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">library</span>(furrr)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>future<span class="sc">::</span><span class="fu">plan</span>(multicore, <span class="at">workers =</span> <span class="dv">4</span>)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>r_3 <span class="ot">=</span> <span class="fu">ici_kendalltau</span>(matrix_2)</span></code></pre></div>
<h2 id="many-many-comparisons">Many Many Comparisons</h2>
<p>In the case of hundreds of thousands of comparisons to be done, the
result matrices can become very, very large, and require lots of memory
for storage. They are also inefficient, as both the lower and upper
triangular components are stored. An alternative storage format is as a
<code>data.frame</code>, where there is a single row for each comparison
performed. This is actually how the results are stored internally, and
then they are converted to a matrix form if requested (the default). To
keep the <code>data.frame</code> output, add the argument
<code>return_matrix=FALSE</code> to the call of
<code>ici_kendalltau</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>r_4 <span class="ot">=</span> <span class="fu">ici_kendalltau</span>(matrix_2, <span class="at">return_matrix =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>r_4</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co">#&gt; $cor</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co">#&gt;   s1 s2 core       raw pvalue   taumax completeness       cor</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co">#&gt; 1 s3 s4    1 0.9924359      0 0.997963        0.921 0.9944616</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co">#&gt; 2 s3 s3    0 1.0000000      0 1.000000        0.950 1.0000000</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co">#&gt; 3 s4 s4    0 1.0000000      0 1.000000        0.950 1.0000000</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co">#&gt; $run_time</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co">#&gt; [1] 0.01733327</span></span></code></pre></div>
<h2 id="other-correlations">Other Correlations</h2>
<p><code>ici_kendalltau</code> and <code>ici_kt</code> calculate the
p-value of the correlation as part of the overall calculation.
<code>stats::cor</code> does not, and <code>stats::cor.test</code> can
only calculate the p-value for a single comparison of two vectors. It is
sometimes advantageous to obtain p-values for a large number of
correlations. We provide <code>cor_fast</code>, which works analogously
to <code>kt_fast</code>, with the ability to choose <code>pearson</code>
or <code>spearman</code> as the method. Note that if a matrix is
provided, the columns must be named.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>r_5 <span class="ot">=</span> <span class="fu">cor_fast</span>(x, y, <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>r_5</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt; $rho</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt;            x          y</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt; x 1.00000000 0.00720612</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt; y 0.00720612 1.00000000</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt; $pvalue</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt;           x         y</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt; x 0.0000000 0.8199608</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt; y 0.8199608 0.0000000</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt; $run_time</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; [1] 0.02877402</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>m_3 <span class="ot">=</span> <span class="fu">cbind</span>(x, y, x)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">colnames</span>(m_3) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;s1&quot;</span>, <span class="st">&quot;s2&quot;</span>, <span class="st">&quot;s3&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>r_6 <span class="ot">=</span> <span class="fu">cor_fast</span>(m_3)</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>r_6</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; $rho</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt;            s1         s2         s3</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt; s1 1.00000000 0.00720612 1.00000000</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt; s2 0.00720612 1.00000000 0.00720612</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; s3 1.00000000 0.00720612 1.00000000</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt; $pvalue</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt;           s1        s2        s3</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt; s1 0.0000000 0.8199608 0.0000000</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="co">#&gt; s2 0.8199608 0.0000000 0.8199608</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="co">#&gt; s3 0.0000000 0.8199608 0.0000000</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="co">#&gt; $run_time</span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co">#&gt; [1] 0.02604246</span></span></code></pre></div>
<h2 id="code-of-conduct">Code of Conduct</h2>
<p>Please note that the ICIKendallTau project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html">Contributor
Code of Conduct</a>. By contributing to this project, you agree to abide
by its terms.</p>

</body>
</html>
