<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Provides functions for calculating
        information-content-informed Kendall-tau. This version of
        Kendall-tau allows for the inclusion of missing values.">
<title>Calculates information-content-informed Kendall-tau • ICIKendallTau</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Calculates information-content-informed Kendall-tau">
<meta property="og:description" content="Provides functions for calculating
        information-content-informed Kendall-tau. This version of
        Kendall-tau allows for the inclusion of missing values.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">ICIKendallTau</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/ici-kendalltau.html">Information-Content-Informed Kendall Tau Correlation</a>
    <a class="dropdown-item" href="articles/testing-for-left-censorship.html">Testing for Left Censorship</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moseleybioinformaticslab/ICIKendallTau/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="icikendalltau">ICIKendallTau<a class="anchor" aria-label="anchor" href="#icikendalltau"></a>
</h1></div>
<!-- badges: start -->

<p>You can see the pkgdown site <a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/" class="external-link">here</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the current version of ICIKendallTau via GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"MoseleyBioinformaticsLab/ICIKendallTau"</span><span class="op">)</span></span></code></pre></div>
<p>You can also install Windows or Mac binaries using our r-universe:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    moseleybioinformaticslab <span class="op">=</span> <span class="st">'https://moseleybioinformaticslab.r-universe.dev'</span>,</span>
<span>    CRAN <span class="op">=</span> <span class="st">"https://cloud.r-project.org"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ICIKendallTau"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="problem">Problem<a class="anchor" aria-label="anchor" href="#problem"></a>
</h2>
<ul>
<li>How to handle missing data (i.e. <code>NA</code>’s) in calculating a correlation between two variables.</li>
<li>Current calculations of correlation are based on having all pairs of observations for two variables.
<ul>
<li>However, whether an observation is present or missing is semi-quantitative information for many analytical measurements with sensitivity limits.</li>
<li>i.e. in many cases, missing observations are not “missing-at-random”, but “missing-not-at-random” due to falling below the detection limit.</li>
<li>In these cases, NA is informative.</li>
<li>Therefore, in <strong>most</strong> analytical measurements (gene expression, proteomics, metabolomics), missing measurements should be included, and contribute to the correlation.</li>
</ul>
</li>
</ul>
<p>If you want to read more on <strong>how</strong> we solve this problem, see the package <a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/articles/ici-kendalltau.html" class="external-link">vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="package-functions">Package Functions<a class="anchor" aria-label="anchor" href="#package-functions"></a>
</h2>
<p>The functions that implement this include:</p>
<ul>
<li>
<code>ici_kt</code>: the C++ workhorse, actually calculating a correlation between an X and Y.
<ul>
<li>The option <code>perspective</code> will control how the <code>NA</code> values influence ties.</li>
<li>When comparing samples, you likely want to use <code>perspective = "global"</code>.</li>
</ul>
</li>
<li>
<code>ici_kendallt</code>: Handles comparisons for a large matrix.
<ul>
<li>Rows are features, columns are samples.</li>
<li>Implicitly parallel, but have to call:
<ul>
<li><code><a href="https://github.com/DavisVaughan/furrr" class="external-link">library(furrr)</a></code></li>
<li><code>plan(multiprocess)</code></li>
</ul>
</li>
<li>Otherwise will only use a single core.</li>
</ul>
</li>
</ul>
<p>We’ve also included a function for testing if the missingness in your data comes from left-censorship, <code>test_left_censorship</code>. We walk through creating example data and testing it in the vignette <a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/articles/testing-for-left-censorship" class="external-link">Testing for Left Censorship</a>. In addition to testing, you can also visualize the missing data pattern by feature rank using the <code>rank_order_data</code> function, and use <code><a href="https://docs.ropensci.org/visdat/reference/vis_miss.html" class="external-link">visdat::vis_miss()</a></code> on the original and reordered missing data.</p>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>The most common case is a large matrix of independent samples (columns) and measured features in each of the samples (i.e. gene expression).</p>
<p>Here we will make some artificial data to show how the correlation changes as we introduce missingness.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://moseleybioinformaticslab.github.io/ICIKendallTau/" class="external-link">ICIKendallTau</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">s1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>, mean <span class="op">=</span> <span class="fl">100</span>, sd <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">s2</span> <span class="op">=</span> <span class="va">s1</span> <span class="op">+</span> <span class="fl">10</span> </span>
<span></span>
<span><span class="va">matrix_1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">s1</span>, <span class="va">s2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">r_1</span> <span class="op">=</span> <span class="fu"><a href="reference/ici_kendalltau.html">ici_kendalltau</a></span><span class="op">(</span><span class="va">matrix_1</span><span class="op">)</span></span>
<span><span class="va">r_1</span><span class="op">$</span><span class="va">cor</span></span>
<span><span class="co">#&gt;    s1 s2</span></span>
<span><span class="co">#&gt; s1  1  1</span></span>
<span><span class="co">#&gt; s2  1  1</span></span></code></pre></div>
<p>Now we introduce some missing values at the low end of each one. We will just do the simplest thing and introduce <code>NA</code> values in the bottom set.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s3</span> <span class="op">=</span> <span class="va">s1</span></span>
<span><span class="va">s3</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">50</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="va">s4</span> <span class="op">=</span> <span class="va">s2</span></span>
<span><span class="va">s4</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">50</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="va">matrix_2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">s3</span>, <span class="va">s4</span><span class="op">)</span></span>
<span><span class="va">r_2</span> <span class="op">=</span> <span class="fu"><a href="reference/ici_kendalltau.html">ici_kendalltau</a></span><span class="op">(</span><span class="va">matrix_2</span><span class="op">)</span></span>
<span><span class="va">r_2</span><span class="op">$</span><span class="va">cor</span></span>
<span><span class="co">#&gt;           s3        s4</span></span>
<span><span class="co">#&gt; s3 1.0000000 0.9944616</span></span>
<span><span class="co">#&gt; s4 0.9944616 1.0000000</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="is-it-fast">Is It Fast?<a class="anchor" aria-label="anchor" href="#is-it-fast"></a>
</h2>
<p>The C++ code implementation (thanks {Rcpp}!) is based on the SciPy implementation, which uses two merge sorts of the ranks of each vector, and then looks for differences between them. This is the fastest method we know of, and has a complexity of O(nlogn). The naive way of computing it, which explicitly examines all of the pairs, has a complexity of n^2. Our implementation was compared to the {pcaPP::cov.fk} function, and the use of {Rcpp} and our inefficient copying of vectors makes ours 3X slower than that one. Which honestly isn’t too bad.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshuaulrich/microbenchmark/" class="external-link">microbenchmark</a></span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">40000</span><span class="op">)</span></span>
<span><span class="va">y2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">40000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html" class="external-link">microbenchmark</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"kendall"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="reference/ici_kt.html">ici_kt</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="st">"global"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="reference/ici_kt.html">ici_kt</a></span><span class="op">(</span><span class="va">x2</span>, <span class="va">y2</span>, <span class="st">"global"</span><span class="op">)</span>,</span>
<span>  times <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Unit: microseconds</span></span>
<span><span class="co">#&gt;                           expr       min        lq      mean    median        uq       max neval</span></span>
<span><span class="co">#&gt;  cor(x, y, method = "kendall") 11755.570 11928.248 12437.300 12409.265 12841.010 13252.408     5</span></span>
<span><span class="co">#&gt;         ici_kt(x, y, "global")   239.267   248.566   413.841   263.888   269.249  1048.235     5</span></span>
<span><span class="co">#&gt;       ici_kt(x2, y2, "global") 13897.996 13991.561 15903.593 15191.136 18077.771 18359.500     5</span></span></code></pre></div>
<p>In the case of 40,000 features, the average time on a modern CPU is 14 milliseconds.</p>
<p>Of course, if you want to use it to calculate Kendall-tau-b without incorporating missingness, it can do that just fine as well.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k_tau</span> <span class="op">=</span> <span class="fu"><a href="reference/ici_kt.html">ici_kt</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">k_tau</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> ,<span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"kendall"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>We also provide the <code>kt_fast</code> function, if you want something that treats <code>NA</code> values similarly to <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">stats::cor</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k_tau_fast</span> <span class="op">=</span> <span class="fu"><a href="reference/kt_fast.html">kt_fast</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span></span>
<span><span class="va">k_tau_fast</span></span>
<span><span class="co">#&gt; $tau</span></span>
<span><span class="co">#&gt;              x            y</span></span>
<span><span class="co">#&gt; x  1.000000000 -0.003411411</span></span>
<span><span class="co">#&gt; y -0.003411411  1.000000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt;           x         y</span></span>
<span><span class="co">#&gt; x 0.0000000 0.8716723</span></span>
<span><span class="co">#&gt; y 0.8716723 0.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $run_time</span></span>
<span><span class="co">#&gt; [1] 0.02890801</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="parallelism">Parallelism<a class="anchor" aria-label="anchor" href="#parallelism"></a>
</h2>
<p>If you have {future} and the {furrr} packages installed, then it is also possible to split up the a set of matrix comparisons across compute resources for any multiprocessing engine registered with {future}.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisVaughan/furrr" class="external-link">furrr</a></span><span class="op">)</span></span>
<span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multicore</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">r_3</span> <span class="op">=</span> <span class="fu"><a href="reference/ici_kendalltau.html">ici_kendalltau</a></span><span class="op">(</span><span class="va">matrix_2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="many-many-comparisons">Many Many Comparisons<a class="anchor" aria-label="anchor" href="#many-many-comparisons"></a>
</h2>
<p>In the case of hundreds of thousands of comparisons to be done, the result matrices can become very, very large, and require lots of memory for storage. They are also inefficient, as both the lower and upper triangular components are stored. An alternative storage format is as a <code>data.frame</code>, where there is a single row for each comparison performed. This is actually how the results are stored internally, and then they are converted to a matrix form if requested (the default). To keep the <code>data.frame</code> output, add the argument <code>return_matrix=FALSE</code> to the call of <code>ici_kendalltau</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r_4</span> <span class="op">=</span> <span class="fu"><a href="reference/ici_kendalltau.html">ici_kendalltau</a></span><span class="op">(</span><span class="va">matrix_2</span>, return_matrix <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">r_4</span></span>
<span><span class="co">#&gt; $cor</span></span>
<span><span class="co">#&gt;   s1 s2 core       raw pvalue   taumax       cor</span></span>
<span><span class="co">#&gt; 1 s3 s4    1 0.9924359      0 0.997963 0.9944616</span></span>
<span><span class="co">#&gt; 2 s3 s3    0 1.0000000      0 1.000000 1.0000000</span></span>
<span><span class="co">#&gt; 3 s4 s4    0 1.0000000      0 1.000000 1.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $run_time</span></span>
<span><span class="co">#&gt; [1] 0.02034473</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-correlations">Other Correlations<a class="anchor" aria-label="anchor" href="#other-correlations"></a>
</h2>
<p><code>ici_kendalltau</code> and <code>ici_kt</code> calculate the p-value of the correlation as part of the overall calculation. <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">stats::cor</a></code> does not, and <code><a href="https://rdrr.io/r/stats/cor.test.html" class="external-link">stats::cor.test</a></code> can only calculate the p-value for a single comparison of two vectors. It is sometimes advantageous to obtain p-values for a large number of correlations. We provide <code>cor_fast</code>, which works analogously to <code>kt_fast</code>, with the ability to choose <code>pearson</code> or <code>spearman</code> as the method. Note that if a matrix is provided, the columns must be named.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r_5</span> <span class="op">=</span> <span class="fu"><a href="reference/cor_fast.html">cor_fast</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span>
<span><span class="va">r_5</span></span>
<span><span class="co">#&gt; $rho</span></span>
<span><span class="co">#&gt;            x          y</span></span>
<span><span class="co">#&gt; x 1.00000000 0.00720612</span></span>
<span><span class="co">#&gt; y 0.00720612 1.00000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt;           x         y</span></span>
<span><span class="co">#&gt; x 0.0000000 0.8199608</span></span>
<span><span class="co">#&gt; y 0.8199608 0.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $run_time</span></span>
<span><span class="co">#&gt; [1] 0.02450991</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">m_3</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s1"</span>, <span class="st">"s2"</span>, <span class="st">"s3"</span><span class="op">)</span></span>
<span><span class="va">r_6</span> <span class="op">=</span> <span class="fu"><a href="reference/cor_fast.html">cor_fast</a></span><span class="op">(</span><span class="va">m_3</span><span class="op">)</span></span>
<span><span class="va">r_6</span></span>
<span><span class="co">#&gt; $rho</span></span>
<span><span class="co">#&gt;            s1         s2         s3</span></span>
<span><span class="co">#&gt; s1 1.00000000 0.00720612 1.00000000</span></span>
<span><span class="co">#&gt; s2 0.00720612 1.00000000 0.00720612</span></span>
<span><span class="co">#&gt; s3 1.00000000 0.00720612 1.00000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt;           s1        s2        s3</span></span>
<span><span class="co">#&gt; s1 0.0000000 0.8199608 0.0000000</span></span>
<span><span class="co">#&gt; s2 0.8199608 0.0000000 0.8199608</span></span>
<span><span class="co">#&gt; s3 0.0000000 0.8199608 0.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $run_time</span></span>
<span><span class="co">#&gt; [1] 0.0231781</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the ICIKendallTau project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/moseleybioinformaticslab/ICIKendallTau/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/moseleybioinformaticslab/ICIKendallTau/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ICIKendallTau</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Robert M Flight <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8141-7788" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Hunter NB Moseley <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3995-5368" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://moseleybioinformaticslab.r-universe.dev" class="external-link"><img src="https://moseleybioinformaticslab.r-universe.dev/badges/ICIKendallTau" alt="ICIKendallTau status badge"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Robert M Flight, Hunter NB Moseley.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
